### PR TITLE
TASK-4-create-components

### PR DESCRIPTION
**Pull request by bombenko**

### CODE REVIEW

File: src/components/component.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    const elem = document.createElement(this.componentName);
+    elem.innerHTML = this.getTemplate();
+    elem.setAttribute(this.hash, '');
+    elem.className += this.componentName;
+
+    return elem;
+  }
+
+  render() {
+    document.querySelector(`[${this.hash}]`).replaceWith(this.getDomElem());
```
Comment (grachpower) [inline src/components/component.js]: move elementRef creation before render

File: src/components/component.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { ComponentService } from '../services/component.service';
+
+export class Component {
+  constructor() {
+    this.hash = new ComponentService().generateHash();
```
Comment (grachpower) [inline src/components/component.js]: predefine componentService, and then use its methods

**Review by grachpower — APPROVED** [review]

File: src/components/smart/card.component.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    super();
+    this.componentName = 'card';
+    this.name = name;
+  }
+
+  /**
+   *
+   * @param {Task} task
+   */
+  deleteListItem(task) {
```
Comment (grachpower) [inline src/components/smart/card.component.js]: return function in closure

File: src/components/smart/card.component.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+   */
+  constructor(name) {
+    super();
+    this.componentName = 'card';
+    this.name = name;
+  }
+
+  /**
+   *
+   * @param {Task} task
```
Comment (grachpower) [inline src/components/smart/card.component.js]: @returns {void}

File: src/components/smart/card.component.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+   */
+  deleteListItem(task) {
+    this.childList = this.childList.filter(elem => elem !== task);
+  }
+
+  getTemplate() {
```
Comment (grachpower) [inline src/components/smart/card.component.js]: api doc

File: src/components/smart/task.component.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+   * @param {string} description
+   */
+  editTask(title, description) {
+    this.title = title;
+    this.description = description;
+  }
+
+  getTemplate() {
+    return `
+      <input class="checkbox" id="${`${this.hash}-cbx`}" type="checkbox">
```
Comment (grachpower) [inline src/components/smart/task.component.js]: find another way to define controls

File: src/services/component.service.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+let service = null;
+
+export class ComponentService {
+  constructor() {
+    if (!service) {
+      service = this;
+    }
+
+    this.componentNum = 0;
```
Comment (grachpower) [inline src/services/component.service.js]: what is that magic number?

File: src/components/component.js (hunk 2 above)
Comment (grachpower) [reply src/components/component.js]: done

File: src/components/component.js (hunk 1 above)
Comment (bombenko) [reply src/components/component.js]: done

File: src/components/component.js (hunk 2 above)
Comment (bombenko) [reply src/components/component.js]: done

File: src/components/smart/card.component.js (hunk 5 above)
Comment (bombenko) [reply src/components/smart/card.component.js]: done

File: src/components/smart/card.component.js (hunk 4 above)
Comment (bombenko) [reply src/components/smart/card.component.js]: done

File: src/services/component.service.js (hunk 7 above)
Comment (bombenko) [reply src/services/component.service.js]: done

**Pull request merged**